### ISSUE TITLE
Feature Request: auto as height

### ISSUE THREAD

**Issue by ealvarezk** [drcmda/react-spring#30]
I was reading the documentation and I couldn't find a way to set auto as enter height prop, as you can see in this example:
```
<Transition
    keys={items.map(item => item.key)}
    from={{ opacity: 0, height: 0 }}
    enter={{ opacity: 1, height: 'auto }}
    leave={{ opacity: 0, height: 0 }}>
    {items.map(item => styles => <li style={styles}>{item.text}</li>)}
</Transition>
```
I think that will make easier to add elements as children without worring the height of those elements. 
If it helps you there is another component, that can do that https://github.com/Stanko/react-animate-height

**Comment by drcmda** [comment]
i think it needs to know how high the element is. I remember struggling with this since i began coding for the web, many years ago. There are some solutions today, flexbox, grid, etc. for instance 0fr - 1fr is numerical, hence it can be animated. Otherwise, react-animate-height looks interesting but it's hard to understand what it's doing. If it doesn't blow out the weight of this lib and if it's not too hacky, perhaps it would make sense implementing it, but would need some help - PRs definitively welcome.

**Comment by drcmda** [comment]
Thanks for these links, very interesting! If we can manage to break it down here, what must be done, and how, i'd open to add it wherever it makes sense, perhaps the animated-core or a helper. In the react-motion issue they were talking about first rendering the thing, calculating height, then overwriting 'auto'. I haven't been fooling around much with the dom lately, but couldn't you measure dom-nodes even if they weren't mounted?

Basically, yes, to the issue here, and yes, i'll be happy to add it, but looking for help and something like cooking receipe on how to pull it off (best with the exact dom calls and functions needed).

**Comment by drcmda** [comment]
Fished this off stackoverflow:

```
HTML

<div class="frame">
    <p>some text some text some text some text</p>
</div>

CSS

.frame {
    width: 120px;
    display: none;
}

.offscreen {
    position: fixed !important;
    left: -9999px !important;
    display: inline !important;
}

JavaScript

$('.frame').addClass('offscreen');

alert('The hieght of \'p\' is: ' + $('p').height() + 'px');

$('.frame').removeClass('offscreen');
```

**Comment by techniq** [comment]
@drcmda I was thinking about how integrating a function for this purpose might work before reading your comment, so might be on to something.  I was thinking of an explicit `measure` function you could pass where you either pass the DOM node and get the measurements yourself (using `node. getBoundingClientRect` or `node.clientWidth`/etc or you pass in the dimensions into this function to make it easier (pass in the `box` (top/left/width/heigh, basically the results of `getBoundingClientRect`).

I think the way react-morph approaches it (for shared element transitions) where you apply the `from` and `to` props to "mark" which elements to measure is interesting...

```js
<ReactMorph>
  {({ from, to, fadeIn, go }) => (
    <div>
      <a onClick={() => go(1)}>
        <strong {...from("title")}>ReactMorph 🐛</strong>
        <br />
        <p {...from("description")}>Morphing transitions was never so easy!</p>
      </a>

      <div>
        <h1 {...to("title")}>ReactMorph 🦋</h1>
        <br />
        <h2 {...to("description")}>Morphing transitions was never so easy!</h2>

        <a onClick={() => go(0)} {...fadeIn()}>
          Back
        </a>
      </div>
    </div>
  )}
</ReactMorph>
```

This isn't really what's being asked here, but the technique might be the same.

Some code worth taking a look at
- https://github.com/brunnolou/react-morph/blob/master/src/index.js
- https://github.com/souporserious/react-measure/blob/master/src/get-content-rect.js

**Comment by drcmda** [comment]
Made a first rough draft:

<img width="1680" alt="screen shot 2018-04-14 at 14 50 53" src="https://user-images.githubusercontent.com/2223602/38768388-49890d30-3ff3-11e8-8c1d-07f4728fdfe3.png">

It's super messy but the essentials are there. animated-core "accepts" the `auto` value, meaning it will just channel through without trying to animate anything. Spring.render will render into a portal when it detects that any value is set to auto, once the portal is created it obtains a ref into the dom node, calls clientWidth/height in there - and theoretically now it would have to inject these values and start the actual render process, but i haven't gotten to that part yet.

**Comment by drcmda** [comment]
<img width="1680" alt="screen shot 2018-04-14 at 14 58 16" src="https://user-images.githubusercontent.com/2223602/38768451-4bd1359e-3ff4-11e8-9351-5d2d8c1c5c37.png">

**Comment by techniq** [comment]
@drcmda the tie to react-dom would be a bummer.  I know for VX we are looking into supporting react-native and `<canvas>` along with `<svg>`.  I know the measurement side will likely have to be DOM-specific, but it would be nice if react-spring could stay agnostic / module (although if you don't use 'auto' I don't guess it would break these other environments).

Any way to pull what you have into the `autoHeight`/`measure`/etc function and maybe make it part of `react-spring-dom` or something?

**Comment by drcmda** [comment]
Thought about it some more, i think there might be a more composable way:

```jsx
import { Spring, animated } from 'react-spring'
import MeasureOnce from 'react-measure-once'

// Variant #1
const App = (
    <Spring native from={{ toggle: 0 }} to={{ toggle: 1 }}>
        {props => (
            <MeasureOnce>
                {({ offscreen, height }) => (
                    <animated.div
                        style={{
                            height: offscreen ? 'auto' : props.toggle.interpolate({ range: [0, 1], output: [0, height] }),
                            width: 300,
                            overflow: 'hidden',
                        }}>
                        Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's
                        standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a
                        type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting,
                        remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing
                        Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of
                        Lorem Ipsum
                    </animated.div>
                )}
            </MeasureOnce>
        )}
    </Spring>
)

// Variant #2
const App = (
    <Spring native from={{ toggle: 0 }} to={{ toggle: 1 }} inject={MeasureOnce}>
        {({ offscreen, toggle, height }) => (
            <animated.div
                style={{
                    height: toggle.interpolate({ range: [0, 1], output: [0, offscreen ? 'auto' : height] }),
                    width: 300,
                    overflow: 'hidden',
                }}>
                Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard
                dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen
                book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially
                unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more
                recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum
            </animated.div>
        )}
    </Spring>
)
```

**Comment by drcmda** [comment]
Alright, working nicely!

![auto height](https://user-images.githubusercontent.com/2223602/38772470-e38c4c3c-4037-11e8-83fd-535721e8597a.gif)

This is how a auto-height spring toggle would look like:

```jsx
<Spring native from={{ toggle: 0 }} to={{ toggle: Math.abs(this.state.open) }} inject={MeasureOnce}>
    {({ offscreen, toggle, height }) => (
        <animated.div
            style={{
            height: offscreen ? 'auto' : toggle.interpolate({ range: [0, 1], output: [0, height] }),
```

It's sure a bit more than just saying height: 'auto' and everything works like magic, but then again, magic will never work in all cases (grids, etc.), but this actually could since it's explicit.

The MeasureOnce component, which would be part of the addons isn't much more than this:

```jsx
class MeasureOnce extends React.PureComponent {
    constructor() {
        super()
        this.width = undefined
        this.height = undefined
        this.portal = document.createElement('div')
        this.portal.style.cssText = 'position:static;visibility:hidden;'
        document.body.appendChild(this.portal)
    }
    measure = ref => {
        if (ref) {
            this.height = ref.clientHeight
            this.width = ref.clientWidth
            document.body.removeChild(this.portal)
            this.forceUpdate()
        }
    }
    render() {
        const { children, instance, ...rest } = this.props
        const result = children({ ...rest, offscreen: this.width === undefined, height: this.height, width: this.width })
        return this.width === undefined ? ReactDOM.createPortal(<div ref={this.measure}>{result}</div>, this.portal) : result
    }
}
```

**Comment by techniq** [comment]
@drcmda killing it as always 😄 

All of this could be encapsulated into a ` <Collapse />` component in userland or as a separate addon/module.  

How often is `MeasureOnce` injected?  I'm assuming once per toggle?  I would figure it would need to be so it can re-measure if the children changes size (adds text, reflow layout, etc).

**Comment by drcmda** [comment]
It would be called on every render, basically wrapping it. For the purpose of a Spring it would be enough, because deeper change detection could be done in the Injection class. I'm toying around a bit more meanwhile, and i think there might be a way to do all of this completely automatic, without having to listen to offscreen flags and so on. The injection component would need access to to/from values, then it would render once with the whole state (which includes 'auto') set to "to", then overwrite 'auto' values and let Spring animate.

**Comment by techniq** [comment]
That sounds great (like animation middleware).

**Comment by techniq** [comment]
Something like this should also work for a react morph-like API where the
from and to are calculated (off different elements). Might also work to
pass from/to into the child render function... But needs more thought.

**Comment by drcmda** [comment]
Got it more or less down: https://codesandbox.io/s/pmjomxn60

The middleware looks like this: https://github.com/drcmda/react-spring/blob/inject/src/addons/fixauto.js

**Comment by techniq** [comment]
Looks great. I hope to get some time in the coming weeks to experiment more
(curious if I can translate one of my react-flip-move examples using this
now, and if shared element transitions are possible as well). Thanks again
for all the hard work.

**Comment by techniq** [comment]
@drcmda I had literally a few minutes to play around with this today but wanted to test out the calculation on children change and works great!  https://codesandbox.io/s/yk7o70v0wv

Still need to find time for the other examples (really curious about how these will port: https://codepen.io/techniq/pen/qNrYYb / https://codepen.io/techniq/pen/WxaxaR)  Hopefully I can get some time this week to give it a shot.

**Comment by drcmda** [comment]
Awesome! BTW, i had some idea for this. animated is made of targets, we're using react-dom atm. It could auto-inject this "middleware" for this target. then nothing would be needed, no addons, no manual injects. And Spring wouldn't rely on react-dom, only when it's using that target. 

In the last update i've made some preparations for this. middleware for instance can channel through if it doesn't want to mutate (for instance if fixAuto doesn't find auto-values). In that can spring behaves like it always would, no wrapping, no portals, etc.

**Comment by drcmda** [comment]
@techniq would you like to try out writing a primitive for flip? I have read a blog about it now, but still scratching my head over what it actually does. I guess i've just never had that problem before.

**Comment by techniq** [comment]
@drcmda FLIP itself is mostly for performance reasons so you only use composite step style attributes (and do not perform a layout or paint step) which basically boils down to only using `transform` and `opacity` to get the end result (using scale to change width/height and translate for x/y basically).  You can see more details on which style properties cause which of the steps (layout, render, composite) here - https://csstriggers.com/ but that's the gist.

Since react-spring readers in javascript and not CSS, some of the concepts do not apply here, but generally just calculating the from/to positions dynamically.

I'm by no means an expert (I mostly just tinker/hack in this space on occasion) but I'll see what I can do.  I'm literally swamped with my work load right now (and baseball season just started up for boys so evenings are consumed) but I'd love to help out as much as I can.

My goal is to try to port those codesandbox examples and see how far I can get.  I'm figuring I'll need to write another middleware that instead of just translating 'auto' to a concrete value, calculating `x/y/width/height` for both `from` and `to` (basically an auto for all).

Thanks again for all the awesome work.  It's really exiting work.

Btw, I know Elijah Meeks was asking the other day if React Native was supported, and while I haven't tested myself, I was pretty sure you planned to support it:

![image](https://user-images.githubusercontent.com/177476/38835091-75815d9c-4198-11e8-8a3a-a0d672a09f9e.png)

I'm also curious how `<canvas />` will work with it, as for `vx` we plan to support it as well as svg and react-native (which is WIP)

**Comment by drcmda** [comment]
@techniq animated comes from react-native, it would work - just needs a few touches. `Spring` is tied atm to targets/react-dom. But just loosely because it references some stuff. I think only 3 things are different between the targets: requestAnimationFrame, cancelAnimationFrame, and the element names (div, span, ..., View, Text, ...). In animated you pull the lib like so: `import Animated from 'animated/targets/react-dom'` but i wanted to make it easier. 

If Airbnb devs are interested, they could help out if they wanted, and pretty much get the lib they need. That's actually the reason i made this, because at work i just couldn't cope any longer with what was already there. If they feel the same, and have react-native experience (i don't), it wouldn't mean that much work to get it up and running.

**Comment by techniq** [comment]
Sounds good. I'll pass the word along. Elijah works at Netflix btw (the VX
devs mostly work at Airbnb though, except me)

**Comment by janhartmann** [comment]
Will this work with `Transition` as well?

**Comment by drcmda** [comment]
@janhartmann it should. All primitives basically revolve around `Spring` so when that one can do something everything should be able to do it. I'm pretty sure i tried it, too - but i can't remember which sandbox that was.

**Comment by janhartmann** [comment]
Hmm, ok. If I do this:

```
<section className={classes.root}>
  <Transition
    keys={items.map(item => item.id)}
    from={{ opacity: 0, height: 0 }}
    enter={{ opacity: 1, height: "auto" }}
    leave={{ opacity: 0, height: 0 }}
  >
    {items.map(item => styles => (
      <div style={styles}>
        <FeedItem>
          {this.getContentType(item)}
        </FeedItem>
      </div>
    ))}
  </Transition>
</section>
```

It fails with a ton of errors in console:

```
Uncaught TypeError: Failed to execute 'getComputedStyle' on 'Window': parameter 1 is not of type 'Element'.
    at ref (react-spring.es.js:1125)
    at commitAttachRef (react-dom.development.js:9918)
    at commitAllLifeCycles (react-dom.development.js:11464)
    at HTMLUnknownElement.callCallback (react-dom.development.js:100)
    at Object.invokeGuardedCallbackDev (react-dom.development.js:138)
    at invokeGuardedCallback (react-dom.development.js:187)
    at commitRoot (react-dom.development.js:11594)
    at completeRoot (react-dom.development.js:12502)
    at performWorkOnRoot (react-dom.development.js:12452)
    at performWork (react-dom.development.js:12370)
```

**Comment by drcmda** [comment]
Ok this might be because of some changes i have made recently. getComputedStyle is new. Do you have a sandbox i could inspect, that would be great!

**Comment by janhartmann** [comment]
I just forked this one: https://codesandbox.io/s/x988w9pmxq

**Comment by drcmda** [comment]
Will take a day, since i'm super busy atm, but tomorrow it'll hopefully be fixed. In the meantime, if it's very pressing, it will most likely start working if you go back a little, something around 5.x.x.

**Comment by janhartmann** [comment]
Dont worry, I am in no hurry! Take your time - I understand open source can be tough together with other projects. :-)

**Comment by drcmda** [comment]
@janhartmann should be fixed in 5.1.8

**Comment by epilande** [comment]
The `fixAuto` doesn't work with `styled-components` probably due to the way `styled-components` handle refs (i.e. innerRef). I forked the code sandbox and added a `styled-components` example https://codesandbox.io/s/mzm40lw178

My current work around is to wrap the `styled` component with a `<div>` element.

**Comment by drcmda** [comment]
@epilande Oh my, i see, this is going to be more complex ... I made a fix in 5.1.9

Your examples started working, i saw a 3 column grid with the elements side by side. If you run into styling issues or wrong height/width, consider this: https://github.com/drcmda/react-spring/blob/master/API-OVERVIEW.md#animating-auto

**Comment by idesignpixels** [comment]
Just going to throw this in here for anyone wanting a simpler solution, you can animate `max-height` or in js `maxHeight` from 0 to a number bigger than you'd need, this way the height animates and it will not exceed its auto height.

**Comment by aleclarson** [comment]
@idesignpixels Nice trick! That will definitely be in the docs for v9. :)
